### PR TITLE
Downloader Login dialog

### PR DESCRIPTION
**Pull request by defvs**
No more connect.sid ! Just type your username/password and it will retrieve it

### CODE REVIEW

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 1]
```diff
@@ -112,13 +139,15 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+
+		private lateinit var cookieStore: BasicCookieStore
 		
 		init {
 			checkConnectsid(CONNECTSID())
 			CONNECTSID.listen { updateConnectsid(it) }
 		}
 		
-		fun execute(httpGet: HttpGet): CloseableHttpResponse {
+		fun execute(httpGet: HttpUriRequest): CloseableHttpResponse {
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]:
```suggestion
		fun execute(request: HttpUriRequest): CloseableHttpResponse {
```

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 2]
```diff
@@ -143,7 +172,7 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
 					domain = "connect.monstercat.com"
 					path = "/"
 					BasicCookieStore().also { it.addCookie(this) }
-				})
+				}.apply { cookieStore = this })
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: I would like to keep this method pure, i.e. no side effects.

File: src/main/xerus/monstercat/downloader/TabDownloader.kt [hunk 3]
```diff
@@ -347,7 +347,7 @@ class TabDownloader: VTab() {
 				show()
 			}
 		}, Button("Checking connection...").apply {
-			prefWidth = 200.0
+			prefWidth = Control.USE_COMPUTED_SIZE
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/downloader/TabDownloader.kt]: The button shouldn't change in size when you interact with it. Why did you change this? Also, if it should use the computed size you can simply remove the line ^^

File: src/main/xerus/monstercat/downloader/TabDownloader.kt [hunk 4]
```diff
@@ -388,18 +395,54 @@ class TabDownloader: VTab() {
...
+					stage.sizeToScene()
+				}
+			}
+			addButton("Cancel") {
+				stage.close()
+			}
+		}
+		parent.children.addAll(emailField, passwordField, controls)
+		stage.isAlwaysOnTop = true
+		stage.focusedProperty().addListener { _, onHidden, _ ->
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/downloader/TabDownloader.kt]: Please reformat this file properly

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 5]
```diff
@@ -74,23 +74,50 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+	fun postLogin(username: String, password: String) =
+		post(HttpPost(uri).apply {
+			setHeader("Accept", "application/json")
+			setHeader("Content-type", "application/json")
+			entity = StringEntity("{\"email\":\"$username\",\"password\":\"$password\"}")
+		})
+	
+	fun postLogout() =
+		post(HttpPost(uri).apply {
+			setHeader("Accept", "application/json")
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: please set these headers for every request and remove this method

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 6]
```diff
@@ -74,23 +74,50 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
-	fun execute() {
+	fun get() {
 		httpGet = HttpGet(uri)
 		response = execute(httpGet!!)
 	}
 	
+	private var httpPost: HttpPost? = null
+	fun post(request : HttpPost) {
+		httpPost = request
+		response = execute(httpPost!!)
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]:
```suggestion
		response = execute(request)
```
that way you don't have to use `!!`, these should be avoided if possible

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 7]
```diff
@@ -74,23 +74,50 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
 	fun getTracks() =
 		parseJSON(TrackResponse::class.java)?.results
 	
+	fun postLogin(username: String, password: String) =
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: Please make this a static method. If you hardcode the JSON you might as well hardcode the URL

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 8]
```diff
@@ -74,23 +74,50 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+		post(HttpPost(uri).apply {
+			setHeader("Accept", "application/json")
+			setHeader("Content-type", "application/json")
+			entity = StringEntity("{\"email\":\"$username\",\"password\":\"$password\"}")
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]:
```suggestion
			entity = StringEntity("""{"email":"$username","password":"$password"}""")
```

Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Wouldn't this not work because $variable doesn't work in plain strings ?

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 2 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: How could I get the cookieStore then ?

File: src/main/xerus/monstercat/downloader/TabDownloader.kt (hunk 3 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/TabDownloader.kt]: I use Computed Size because the text inside the button changes and 200 wasn't enough. I find it better to have the button exactly to it's text size.

I could simply make prefWidth higher.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 5 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: This should be removed indeed; to logout all we need to do is remove the connect.sid

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 8 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Nevermind just checkeed the doc it does work.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 2 above)
Comment (Xerus2000) [reply src/main/xerus/monstercat/api/APIConnection.kt]: I'm sure you can find a solution ;) otherwise you have to split it up or something. But a method starting with `create` should not modify anything, solely create.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 5 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Removed in a81ff488b772515fd42a35feb308eb07e3cb503e

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 7 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Refactored in a81ff488b772515fd42a35feb308eb07e3cb503e

File: src/main/xerus/monstercat/downloader/TabDownloader.kt (hunk 3 above)
Comment (Xerus2000) [reply src/main/xerus/monstercat/downloader/TabDownloader.kt]: Then do that. In my experience it looks really weird if the button resizes automatically, and it can lead to annoying interaction issues.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 2 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Using HttpClientContext here : 6cc64ba6d2ca7d3e42403dbefb28b93df5cd7f29

File: src/main/xerus/monstercat/downloader/TabDownloader.kt (hunk 4 above)
Comment (defvs) [reply src/main/xerus/monstercat/downloader/TabDownloader.kt]: f3800ca3a21844a4ca20fc8d25aea86c4095ca04

**Comment by defvs** [comment]
Branch ready, if you have any other request, feel free to comment.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 2 above)
Comment (Xerus2000) [reply src/main/xerus/monstercat/api/APIConnection.kt]: have you tested this? But it definitely looks better to remove this class-wide variable!

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 9]
```diff
@@ -77,20 +78,34 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
 	
 	private var httpGet: HttpGet? = null
-	fun execute() {
+	fun get(context: HttpClientContext? = null) {
 		httpGet = HttpGet(uri)
-		response = execute(httpGet!!)
+		response = execute(httpGet!!, context)
+	}
+	
+	private var httpPost: HttpPost? = null
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: I don't think you need 3 separate variables for get, post and put, one should suffice since it's just for aborting the request

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 10]
```diff
@@ -200,11 +215,34 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+			val connection = APIConnection("v2", "signin")
+			val context = HttpClientContext()
+			connection.post(HttpPost(connection.uri).apply {
+				setHeader("Accept", "application/json")
+				setHeader("Content-type", "application/json")
+				entity = StringEntity("""{"email":"$username","password":"$password"}""")
+			}, context)
+
+			val code = connection.response?.statusLine?.statusCode ?: 0
+			logger.debug("code returned is $code")
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: Either make a proper debug message or remove it

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 11]
```diff
@@ -200,11 +215,34 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+		fun login(username: String, password: String): Boolean {
+			val connection = APIConnection("v2", "signin")
+			val context = HttpClientContext()
+			connection.post(HttpPost(connection.uri).apply {
+				setHeader("Accept", "application/json")
+				setHeader("Content-type", "application/json")
+				entity = StringEntity("""{"email":"$username","password":"$password"}""")
+			}, context)
+
+			val code = connection.response?.statusLine?.statusCode ?: 0
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: why the "?: 0" ?

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 12]
```diff
@@ -200,11 +215,34 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
+			logger.debug("code returned is $code")
+			if (code !in 200..206)
+				return false
+			val connectsid = context.cookieStore.cookies.find { it.name == "connect.sid" }?.value ?: return false
+			logger.debug("connectsid is $connectsid")
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: this will already be logged because of the Settings update afaik

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 2 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Yes, works great.

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 11 above)
Comment (Xerus2000) [reply src/main/xerus/monstercat/api/APIConnection.kt]: but why not leave it as `null`, to make it clear there was no value?

Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: it's an int ? and we already do a check on whether the response code is on the 200-OK range, so 0 never passes

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 9 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: b2a0cdfb6c79f566c4440650de445f2f5bd29baf

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 10 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Made better AND trace with 94d29ef1915da2c73fe8451d994bf5e47c56b101

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 11 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Removed it. It is an int, so if it is null, it isn't in the 200 range. 94d29ef1915da2c73fe8451d994bf5e47c56b101

File: src/main/xerus/monstercat/api/APIConnection.kt (hunk 12 above)
Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: Removed in 94d29ef1915da2c73fe8451d994bf5e47c56b101

File: src/main/xerus/monstercat/api/APIConnection.kt [hunk 13]
```diff
@@ -73,24 +74,34 @@ class APIConnection(vararg path: String) : HTTPQuery<APIConnection>() {
...
-	fun execute() {
-		httpGet = HttpGet(uri)
-		response = execute(httpGet!!)
+
+	fun get(context: HttpClientContext? = null) {
+		httpRequest = HttpGet(uri)
+		response = execute(httpRequest!!, context)
+	}
+
+	fun post(request : HttpPost, context: HttpClientContext? = null) {
```
Comment (Xerus2000) [inline src/main/xerus/monstercat/api/APIConnection.kt]: Why do we even have different methods for this? I think there could be one which simply takes `request: HttpUriRequest`, just called `execute`, and it could be used by the `get` method as well

Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: 05fd56b27c85faf3947adf010501a57e1564cf1a

Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: One execute() for everything. Just add your own HttpGet / HttpPost / HttpPut as argument to choose which request type to use.

Comment (defvs) [reply src/main/xerus/monstercat/api/APIConnection.kt]: I renamed the old execute in the Companion because it had the same arguments as the new function.
